### ISSUE TITLE
zramSwap.enable has no effect on boot (easily reproduced with nixos-rebuild build-vm)

### ISSUE THREAD

**Issue by falsifian** [NixOS/nixpkgs#86061]
**Describe the bug**
I have `zramSwap.enable` set to `true`, but zram is not enabled on boot.

I originally reported it in [this discourse thread](https://discourse.nixos.org/t/zramswap-enable-true-does-nothing/6734?u=falsifian); there are some notes there about how I'm able to manually get zram started after boot. The steps below show how to reproduce the issue with `nixos-rebuild build-vm`.

**To Reproduce**
1. Check out nixpkgs somewhere. On my computer it's at ~/bin/work/nixpkgs. I have tried this with `master` (commit 5e4abf76) and `release-20.03` (commit c0c339b6).
2. Write a `configuration.nix` file with the below contents. I put mine at `~/tmp/configuration.nix`.
3. `NIX_PATH=nixpkgs=$HOME/big/work/nixpkgs:nixos=$HOME/big/work/nixpkgs NIXOS_CONFIG=$HOME/tmp/configuration.nix nixos-rebuild --no-build-output build-vm`
4. `result/bin/run-nixos-vm`
5. In the vm, log in. Type `free -m`; observe "0 0 0" on the "swap:" line. `ls /dev/zram0` says it does not exist.

configuration.nix

    { ... }:

    {
      users.users = {
        falsifian = {
          password = "falsifian";
          isNormalUser = true;
          extraGroups = [ "wheel" ];
        };
      };
      zramSwap = {
        enable = true;
      };
    }

**Expected behavior**
As documented, I expect `zramSwap.enable` to cause zram swap to be enabled on boot.

**Notify maintainers**
(from git log) @danbst @wmertens 

**Metadata**
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

     - system: `"x86_64-linux"`
     - host os: `Linux 5.4.33, NixOS, 20.03.1445.95b9c99f6d0 (Markhor)`
     - multi-user?: `yes`
     - sandbox: `yes`
     - version: `nix-env (Nix) 2.3.4`
     - channels(james): `""`
     - channels(root): `"nixos-20.03.1445.95b9c99f6d0"`
     - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute:
# a list of nixos modules affected by the problem
module:
```

**Comment by bandresen** [comment]
Can you try to add
```
swapDevices = [ ]
```

to your configuration.nix to see if that helps?

**Comment by falsifian** [comment]
@bandresen It helps with my real configuration, but not with the vm configuration.

Adding `swapDevices = [ ];` to the end of the vm configuration does not change the store path.

When I went to add it to my real configuration, I noticed I had a `mkForce` for `swapDevices`, as a hack to deal with bad behaviour by nixos-generate-config. (My configuration is in the discourse thread.) Simply commenting out my `swapDevices` attribute was enough to make zram be set up on boot. (For completeness, I also tried adding `swapDevices = [ ];` after commenting out my `mkForce` and I don't think the store path changed.)

**Comment by wmertens** [comment]
I think it could have been because of your mkForce. The zram module only adds a service to set up a zram device and adds the device to swapDevices here: https://github.com/NixOS/nixpkgs/blob/62e866e4fd6e414c469f365e1c9b3fb0c80b8e8d/nixos/modules/config/zram.nix#L178

So I don't think it's a bug, but a configuration error?

**Comment by falsifian** [comment]
@wmertens yes, in the case of my real configuration, I think it could be considered an error on my part.

I changed the title of the bug to refer just to the `nixos-rebuild build-vm` case, since I still don't know what causes zram not to (reproducibly) not work in my vm configuration. Perhaps `build-vm` does something magical with the swap configuration?

**Comment by falsifian** [comment]
(To be clear, this is not a real problem for me, but I'm leaving this issue open because it does seem to be a bug as far as I can tell.)

**Comment by wmertens** [comment]
Would it be that hard to keep only the zram devices in the vm definition?

**Comment by tejing1** [comment]
Yes. The nixos module system doesn't allow inspecting the "previous" value of an option "before" overriding it, so you'd have to... um... let's see... I guess introduce a second "real" swap devices option, which normally gets set to a copy of the user-visible one, and have the vm code override that with a filtered copy instead.

**Comment by Artturin** [comment]
zram.nix has been changed a lot and doesn't use swapDevices anymore
